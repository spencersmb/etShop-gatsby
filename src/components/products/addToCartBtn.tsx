import { IState } from '@et/types/State'
import { addProductToCart, cartToggle as cartToggleAction, IAddProductAction } from '../../state/actions/cartActions'
import React from 'react'
import { Action, bindActionCreators, Dispatch } from 'redux'
import { connect } from 'react-redux'
import { ICartState } from '@et/types/Cart'
import { IProduct } from '@et/types/Products'

interface IPropsPublic {
	handleAddToCartState: () => void;
	isInCart: boolean;
	slug: string;
	selectedProduct: IProduct;
	licenseQty: number | string;
	price: string
	total: string
}

interface IPropsReduxActions {
	addToCart: IAddProductAction,
	cartToggle: () => void,
}

interface IPropsPrivate {
	cart: ICartState,
}

export function AddToCartBtn (props: IPropsPublic & IPropsPrivate & IPropsReduxActions) {
	const { addToCart, cart, selectedProduct, handleAddToCartState, licenseQty, price, slug, cartToggle, isInCart, total } = props
	const disabled = (licenseQty === 0) || (typeof licenseQty === 'string')

	async function handleAddToCart () {
		// secure check for 0 licenses
		if (typeof licenseQty !== 'string') {
			handleAddToCartState()
			addToCart(selectedProduct, cart.items, slug, licenseQty, price)
		}
	}

	if (isInCart) {
		return (
			<button data-testid='checkout' onClick={cartToggle}>
				Checkout
			</button>
		)
	}

	return (
		<>
			<button data-testid='addToCart' onClick={handleAddToCart} disabled={disabled}>
				I want this <span>{total}</span>
			</button>
			{disabled && <div data-testid='warning'>Must have at least one computer license selected</div>}
		</>
	)
}

const mapStateToProps = (state: IState): { cart: ICartState } => {
	return {
		cart: state.cart
	}
}

const mapDispatchToProps = (dispatch: Dispatch<Action>) => {
	return {
		addToCart: bindActionCreators(addProductToCart, dispatch),
		cartToggle: bindActionCreators(cartToggleAction, dispatch)
	}
}

export default connect<IPropsPrivate, IPropsReduxActions, IPropsPublic, IState>(mapStateToProps, mapDispatchToProps)(AddToCartBtn)
