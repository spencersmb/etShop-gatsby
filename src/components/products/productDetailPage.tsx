import ProductLayout from '@components/products/productLayout'
import SEO from '@components/seo'
import { IGatsbyConfig } from '@et/types/Gatsby'
import { IProduct } from '@et/types/Products'
import { IMeta, IOGType } from '@et/types/SEO'
import { jsonldImages, socialUtils, twitterDefaultMeta } from '@utils/genUtils'
import { graphql } from 'gatsby'
import React, { Component } from 'react'

type Response = IGatsbyConfig & { wcProduct: IProduct }

interface IProductQuery {
	data: Response
}

// WRAP IN GRAPHQL to pass data to twitter from sitemetadata
export class ProductDetailPage extends Component<IProductQuery> {
	ogArticles: IOGType[] = []
	twitterAddons: IMeta[] = []
	jsonld: any
	price: string

	constructor (props: IProductQuery) {
		super(props)
		const { data: { wcProduct, site: { siteMetadata } } } = this.props
		this.ogArticles = wcProduct.tags.map(tag => tag.name).map(article => ({
			property: 'article:tag',
			content: `${article}`
		}))

		this.price = wcProduct.on_sale ? wcProduct.sale_price : wcProduct.price

		this.twitterAddons = [
			{
				name: `twitter:card`,
				content: `summary_large_image`
			},
			{
				name: `twitter:title`,
				content: `${wcProduct.name}`
			},
			{
				name: `twitter:description`,
				content: `${wcProduct.seo.desc}`
			},
			{
				name: `twitter:image`,
				content: `${wcProduct.images.length > 0 ? wcProduct.images[0].thumbnail.url : socialUtils.twitter.defaultImage}`
			}
		]

		this.jsonld = {
			['@context']: 'http://schema.org/',
			[`@type`]: 'Product',
			image: [
				...jsonldImages(wcProduct.images)
			],
			description: `${wcProduct.seo.desc}`,
			sku: `${wcProduct.product_id}`,
			brand: {
				'@type': 'Thing',
				'name': 'Every-Tuesday'
			},
			offers: {
				['@type']: 'Offer',
				priceCurrency: 'USD',
				priceValidUntil: '2300-02-18T23:26:37+0000',
				itemCondition: 'https://schema.org/NewCondition',
				availability: 'http://schema.org/InStock',
				seller: {
					['@type']: 'Organization',
					name: 'Every-Tuesday'
				},
				price: `${this.price}`
			},
			name: wcProduct.name
		}
	}

	render () {
		// TODO: add featured image to meta - currently the thumbnail
		// TODO: get google verification token

		const { data: { wcProduct, site: { siteMetadata } } } = this.props

		return (
			<>
				<SEO
					title='page title for product 1'
					description={`description for product1`}
					meta={[
						{
							property: `og:type`,
							content: `article`
						},
						{
							property: `og:price:amount`,
							content: this.price
						},
						{
							property: `og:price:currency`,
							content: `USD`
						},
						{
							property: `og:price:availability`,
							content: `instock`
						},
						{
							property: `og:updated_time`,
							content: `${wcProduct.date_modified_gmt}`
						},
						{
							property: `article:publisher`,
							content: `${siteMetadata.siteUrl}`
						},
						{
							property: `article:publishedTime`,
							content: `${wcProduct.date_created_gmt}`
						},
						{
							property: `article:modifiedTime`,
							content: `${wcProduct.date_modified_gmt}`
						},
						{
							property: `article:authors`,
							content: `${siteMetadata.authorUrl}`
						},
						...this.ogArticles,
						// Twitter
						...twitterDefaultMeta(this.twitterAddons)
					]}
				>

					<link rel='canonical' href={`${process.env.GATSBY_DB}/products/${wcProduct.slug}`}/>
					<script type='application/ld+json'>{JSON.stringify(this.jsonld)}</script>
				</SEO>
				<ProductLayout product={wcProduct}/>
			</>
		)
	}
}

export default ProductDetailPage

export const productQuery = graphql`
	query SingleProductQuery($slug: String!){
		site {
			siteMetadata {
				title
			}
		}
		wcProduct(
			slug:{
				eq: $slug
			}
		){
			name
			sub_header
			id
			images{
				thumbnail{
					url
				}
			}
			categories{
				name
			}
			license{
				type
				hasExtendedLicense
				standardItem{
					slug
					bullets{
						bullet_point
					}
				}
				extendedItem{
					slug
					bullets{
						bullet_point
					}
				}
			}
			date_created_gmt
			date_modified_gmt
			price
			product_id
			regular_price
			pwyw
			sale_price
			seo{
				desc
				title
			}
			slug
			tags{
				name
				slug
			}

			type
		}
	}
`