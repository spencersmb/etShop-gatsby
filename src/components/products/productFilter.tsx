import { device } from '@styles/global/breakpoints'
import { colors } from '@styles/global/colors'
import { Sentinel } from '@styles/global/fonts'
import { svgs } from '@svg'
import { renderSvg } from '@utils/styleUtils'
import { getWindowSize } from '@utils/windowUtils'
import React, { Ref, useEffect, useRef, useState } from 'react'
import posed from 'react-pose'
import styled from 'styled-components'

interface IProps {
	handleClick: (filter: string) => void
	filter: string
}

interface IFilterItem {
	slug: string,
	name: string,
	icon: string
}

const filterItems: IFilterItem[] = [
	{
		slug: 'fonts',
		name: 'Fonts',
		icon: 'Fonts'
	},
	{
		slug: 'textures',
		name: 'Textures',
		icon: 'Textures'
	},
	{
		slug: 'templates',
		name: 'Templates',
		icon: 'Templates'
	}
]

function getNavSize (windowSize: string) {

	if (windowSize === 'desktop') {
		return 87
	} else {
		return 75 - 65
	}
}

const useScrollEvent = () => {
	const [fixed, setFixed] = useState(false)
	const [mq, setMq] = useState('')
	const prevFixed = useRef(fixed)
	const prevMQ = useRef(mq)

	useEffect(() => {
		prevFixed.current = fixed
	}, [fixed])

	useEffect(() => {
		prevMQ.current = mq
	}, [mq])

	useEffect(() => {
		const header = document.getElementById('header')
		const cartCheck = document.getElementById('cart-Container')
		let carthasChildren = cartCheck ? cartCheck.children.length > 0 : false
		let headerHeight = 0
		let windowDevice = getWindowSize()
		let size = windowDevice === 'desktop' ? headerHeight + getNavSize(windowDevice) : headerHeight - getNavSize(windowDevice)

		const reqAnim = () => {
			requestAnimationFrame(watchNav)
		}
		const watchNav = () => {
			const fromTop = window.scrollY
			windowDevice = getWindowSize()
			// console.log('windowDevice', windowDevice)
			// console.log('prevMQ.current', prevMQ.current)

			if (prevMQ.current !== windowDevice) {
				carthasChildren = cartCheck ? cartCheck.children.length > 0 : false
				headerHeight = header ? header.getBoundingClientRect().height : 0
				size = windowDevice === 'desktop' ? headerHeight + getNavSize(windowDevice) : headerHeight - getNavSize(windowDevice)
				setMq(windowDevice)
				// console.log('change', windowDevice)
			}

			if (fromTop >= size && !prevFixed.current) {
				// console.log('sticky')
				setFixed(true)

			} else if (fromTop < size && prevFixed.current && carthasChildren) {
				// link.classList.remove("current");
				setFixed(true)
				// console.log('Sticky cart')
			} else if (fromTop < size && prevFixed.current) {
				// link.classList.remove("current");
				setFixed(false)
				// console.log('unStick')
			}
		}

		window.addEventListener('scroll', reqAnim)

		return () => {
			window.removeEventListener('scroll', reqAnim)
		}
	}, [])

	return [fixed]
}

/*
 * Filter Feature - HOW IT WORKS
 * This sticky nav uses a custom useEffect hook to add a window listener on the scroll event to and returns a boolean on if the
 * Nav should be sticky or not based on the headers height. Should be updated to just detect where the nav is on the page. Works for now though.
 *
 * The Element Clicked function checks if the nav is sticky to determine where to scroll back up to after someone clicks a filter.
 * Essentially we want the user to be rescrolled to the top of the newly filtered content so that nothing is cut off when a filter is clicked.
 *
 * Inside it we check the device width to account for the mobile sticky nav height if on tablet or phone.
 * Additionally if the nav is open we close it to show the content after a filter is clicked.
 *
 *
 */
function ProductFilter (props: IProps) {
	const { handleClick } = props
	const [isOpen, setIsOpen] = useState(false)
	const [fixed] = useScrollEvent()
	const filterContainerRef = useRef(null)
	const headerRef: any = useRef<HTMLElement>(null)

	useEffect(() => {
		headerRef.current = document.getElementById('header')
	}, [])

	function elementClick (e: any) {
		e.preventDefault()
		if (fixed && filterContainerRef.current) {
			const y = headerRef.current ? headerRef.current.getBoundingClientRect().height : 0
			const deviceWidth = getWindowSize()
			// nav height used because its added onto a 2nd sticky nav when in mobile else use 0 because there isnt anything else above
			const filterNavHeight = deviceWidth === 'desktop' ? 0 : 74
			const navHeight = deviceWidth === 'desktop' ? 87 : 75
			// console.log('y', y)
			// console.log('window.scrollY', window.scrollY)

			window.scrollTo(0, (y + navHeight) - filterNavHeight)
		}
		if (isOpen) {
			setIsOpen(false)
		}
		handleClick(e.currentTarget.getAttribute('data-filtertype'))
	}

	function handleToggle () {
		setIsOpen(!isOpen)
	}

	return (
		<FilterContainer ref={filterContainerRef} id={'filterContainer'}>
			<Filter fixed={fixed}>
				<FilterInner>
					<FilterHeader onClick={handleToggle}>
					<span>
						{renderSvg(svgs.Filter)}
					</span>
						<h4>
							Filter
						</h4>
					</FilterHeader>
					<FilterList pose={isOpen ? 'open' : 'closed'}>
						{filterItems.map((item: IFilterItem) => (
							<FilterListItem
								key={item.slug}
								selectedFilter={props.filter === item.slug}
								data-testid='filterItems'
								onClick={elementClick}
								data-filtertype={item.slug}
							>
								<FilterContent>
									<Svg selectedFilter={props.filter === item.slug}>
										{renderSvg(svgs[item.icon])}
									</Svg>
									{item.name}
								</FilterContent>
								<Slider className='slider' selectedFilter={props.filter === item.slug}/>
							</FilterListItem>
						))}
						<FilterListItem
							key={'view-all'}
							selectedFilter={props.filter === ''}
							data-testid='filterItems'
							onClick={elementClick}
							data-filtertype={''}
						>
							<FilterContent>
								View all
							</FilterContent>
							<Slider className='slider' selectedFilter={props.filter === ''}/>
						</FilterListItem>
					</FilterList>
					<FilterViewAll
						data-testid='filterItems'
						onClick={elementClick}
						selectedFilter={props.filter === ''}
						data-filtertype=''>
						<FilterContent>
							View all
						</FilterContent>
						<Slider className='slider' selectedFilter={props.filter === ''}/>
					</FilterViewAll>
				</FilterInner>
			</Filter>
		</FilterContainer>
	)
}

const ListPosed = posed.ul({
	closed: {
		height: '0px',
		transition: {
			default: {
				ease: 'easeOut'
			}
		}
	},
	open: {
		height: 'auto',
		transition: {
			default: {
				ease: 'backInOut'
			}
		}
	}
})
const FilterList = styled(ListPosed)`
	background-color: ${colors.grey.i400};
	margin:0;
	padding: 0;
	display: flex;
	flex-direction: column;
	justify-content: center;
	overflow: hidden;
	@media ${device.laptop} {
		flex-direction: row;
		height: auto !important;
	}
`
const FilterContainer = styled.div`
	grid-column:1 / -1;
	grid-row: 1;
	z-index: 2;
	position:relative;
	height: 74px;
	//overflow: hidden;
`
const Filter = styled.div<{ fixed: boolean }>`
	position: ${props => props.fixed ? 'fixed' : ''};
	width: 100%;
	top: 75px;
	background: ${colors.grey.i400};
	z-index: 2;
	
	@media ${device.laptop} {
		top: 0;
	}
`
const FilterInner = styled.div`
	max-width: 1200px;
	margin: 0 auto;
	display: flex;
	flex-direction: column;
	width: 100%;

	@media ${device.laptop} {
		flex-direction: row;
		justify-content: space-between;
	}
		
`

interface IFilterListItem {
	selectedFilter: boolean
}

const FilterListItem = styled.li<IFilterListItem>`
	list-style: none;
	display: flex;
	flex-direction: column;
	cursor: pointer;
	padding: 20px 20px 20px;
	margin: 0;
	font-size: 15px;
	font-weight: 500;
	text-transform: uppercase;
	color: ${props => props.selectedFilter ? `#fff` : colors.primary.headline};
	transition: color .3s;
	position: relative;
	background: ${props => props.selectedFilter ? colors.teal.i500 : 'transparent'};
	
	@media ${device.laptop} {
		color: ${props => props.selectedFilter ? colors.teal.i500 : colors.primary.headline};
		background: transparent;
		&:hover{
			color: ${colors.teal.i500};
			border-color: ${colors.teal.i500};
			
			svg{
				path{
					fill: ${colors.teal.i500};
				}
			}
			
			.slider{
				background: ${colors.teal.i500};
				width: 100%;
			}
		}			
		&:last-child{
			display: none;
		}
	}
		
`
const FilterContent = styled.div`
	display: flex;
	flex-direction: row;
	align-items: center;
`
const Svg = styled.span<IFilterListItem>`
	width: 20px;
	height: 20px;
	margin-right: 5px;
	svg{
		path{
			transition: .3s;
			fill: ${props => props.selectedFilter ? `#fff` : colors.primary.headline};
		}
	}
	@media ${device.laptop} {
		svg{
			path{
				transition: .3s;
				fill: ${props => props.selectedFilter ? colors.teal.i500 : colors.primary.headline};
			}
		}
	}
		
`
const Slider = styled.span<IFilterListItem>`
	height: 2px;
	width: ${props => props.selectedFilter ? '100%' : '0%'};
	display: block;
	background: ${colors.teal.i500};
	transition: color .3s, width .4s;
	position: absolute;
	bottom: 0;
	left: 0;
`
const FilterHeader = styled.div`
	padding: 0 20px 0 20px;
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	margin: 13px 0 10px;
	h4{
		${Sentinel.reg};
		font-weight: 600;
		font-style: italic;
		font-size: 20px;
		color: ${colors.primary.headline};
		text-transform: uppercase;
		margin-bottom: 0;
	}
	span{
		width: 22px;
		margin-right: 15px;
		display: flex;
		align-items: center;
	}
	svg{
		width: 100%;
	}
	path{
		fill: ${colors.primary.headline};
	}
	
	@media ${device.laptop} {
		margin: 0;    
	}
		
`
const FilterViewAll = styled(FilterListItem)<IFilterListItem>`
	// padding: 20px;
	// text-transform: uppercase;
	// font-weight: 500;
	// color: ${props => props.selectedFilter ? colors.teal.i500 : colors.primary.headline};
	// cursor: pointer;
	display: none;
	
	
	@media ${device.laptop} {
		display: flex;
		&:last-child{
		display: flex;
		}	    
	}
		
`
export default ProductFilter
