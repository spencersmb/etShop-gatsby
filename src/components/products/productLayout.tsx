import AddToCartBtn from '@components/cart/addToCartBtn'
import NumberDial from '@components/forms/inputs/numberDial'
import LicenseSelect from '@components/forms/inputs/productSelect'
import Layout from '@components/layout'
import { ICartItem, ICartState } from '@et/types/Cart'
import { IProduct, IProducts } from '@et/types/Products'
import { IState } from '@et/types/State'
import { checkCartForProduct } from '@utils/cartUtils'
import { calcBulkPriceDiscount, displayCurrency } from '@utils/priceUtils'
import { Link } from 'gatsby'
import React, { Dispatch, useEffect, useLayoutEffect, useReducer, useRef } from 'react'
import config from '../../../gatsby-config'
import { connect } from 'react-redux'
import styled from 'styled-components'
import _ from 'lodash'

interface IPropsPrivate {
	products: IProducts,
	cart: ICartState
}

interface IPropsPublic {
	product: IProduct
}

interface IPublicState {
	selectedProduct: IProduct,
	selectedLicense: string,
	numberOfLicenses: number | string,
	inCart: boolean,
	bulkDiscount: boolean,
	price: string,
	payWhatYouWant: boolean
}

interface INewState {
	selectedProduct?: IProduct,
	selectedLicense?: string,
	numberOfLicenses?: number | string
	inCart?: boolean,
	bulkDiscount?: boolean,
	price?: string,
	payWhatYouWant?: boolean
}

type useSetStateType = [IPublicState, Dispatch<INewState>]

function useSetState (initialState: any): useSetStateType {
	const [state, setState] = useReducer((originalState: IPublicState, newState: INewState) => ({ ...originalState, ...newState }),
		initialState)

	return [
		state,
		setState
	]
}

export const ProductLayout = (props: IPropsPublic & IPropsPrivate) => {
	const { product, products, cart } = props
	const [state, setState] = useSetState({
		selectedProduct: product,
		selectedLicense: 'standard',
		numberOfLicenses: 1,
		inCart: false,
		bulkDiscount: false,
		payWhatYouWant: product.pwyw,
		price: product.price
	})
	const standardItem = useRef(props.product)
	const extendedItem = useRef(product.license.extendedItem ? products[product.license.extendedItem.slug] : null)

	// dispatch effect after cart is loaded
	// check for product in cart on load
	// then dispatch to state the item found
	useLayoutEffect(() => {
		// Check if item is in cart on load
		const productFound: boolean = checkCartForProduct(cart, product.slug).length > 0

		if (productFound && !state.inCart) {

			// 1. Get the item in the cart
			const cartItem: ICartItem = cart.items[product.slug]

			// 2. Set the correct item
			const selectedProduct = cartItem.extended && extendedItem.current ? extendedItem.current : standardItem.current
			const bulkDiscount: boolean = cartItem.qty >= config.siteMetadata.pricing.minQuantity

			// 3. Set state
			setState({
				selectedProduct,
				selectedLicense: cartItem.extended ? 'extended' : 'standard',
				numberOfLicenses: cartItem.qty,
				inCart: true,
				bulkDiscount,
				payWhatYouWant: selectedProduct.pwyw,
				price: cartItem.price
			})
		}

	}, [cart.loaded])

	// on Cart price change
	// if the item gets removed,
	// reset the pageLayout instance
	useEffect(() => {
		if (previousCart.current.loaded) {
			const productFound: boolean = checkCartForProduct(cart, product.slug).length > 0

			if (!productFound) {
				setState({
					inCart: false,
					numberOfLicenses: 1,
					selectedLicense: 'standard',
					bulkDiscount: false,
					selectedProduct: props.product,
					price: props.product.price
				})
			} else {

				// check if the items are equal
				if (!_.isEqual(previousCart.current.items[product.slug], cart.items[product.slug])) {
					// product is in cart and may have changed
					// 1. Get the item in the cart
					const cartItem: ICartItem = cart.items[product.slug]

					// 2. Set the correct item
					const selectedProduct = cartItem.extended && extendedItem.current ? extendedItem.current : standardItem.current
					const bulkDiscount: boolean = cartItem.qty >= config.siteMetadata.pricing.minQuantity

					// 3. Set state
					setState({
						selectedProduct,
						selectedLicense: cartItem.extended ? 'extended' : 'standard',
						numberOfLicenses: cartItem.qty,
						inCart: true,
						bulkDiscount,
						payWhatYouWant: selectedProduct.pwyw,
						price: cartItem.price
					})
				}

			}
		}
	}, [cart.totalPrice, cart.totalItems])

	const previousCart = useRef(cart)
	useEffect(() => {
		previousCart.current = cart
	})

	function selectChange (e: any) {
		setState({
			selectedLicense: e.target.value
		})

		if (e.target.value === 'standard') {
			setState({
				selectedProduct: standardItem.current,
				price: standardItem.current.on_sale
					? standardItem.current.sale_price
					: calcBulkPriceDiscount(state.bulkDiscount, standardItem.current.price)
			})
		} else if (e.target.value === 'extended' && extendedItem.current) {
			setState({
				selectedProduct: extendedItem.current,
				price: extendedItem.current.on_sale
					? extendedItem.current.sale_price
					: calcBulkPriceDiscount(state.bulkDiscount, extendedItem.current.price)
			})
		}
	}

	function onDialChange (total: number | string) {

		if (total >= config.siteMetadata.pricing.minQuantity) {
			setState({
				numberOfLicenses: total,
				bulkDiscount: true,
				price: calcBulkPriceDiscount(true, state.price)
			})
		} else {
			setState({
				numberOfLicenses: total,
				bulkDiscount: false,
				price: state.selectedProduct.price
			})
		}

	}

	function onPwywChange (total: number | string) {
		setState({
			price: total.toString()
		})

	}

	const { name, license: { hasExtendedLicense } } = props.product

	return (
		<Layout>
			<h1>Item name: {name}</h1>
			{/*Convert to readable price helper fn?*/}
			<h2>price: {displayCurrency(state.price)}</h2>
			<hr/>
			<div>
				<p>State:</p>
				<p>Product: <b>{state.selectedProduct.slug}</b></p>
				<p># of licenses: <b>{state.numberOfLicenses}</b></p>
			</div>
			<hr/>
			<p>has extended license: <span
				style={{ color: hasExtendedLicense ? 'green' : 'blue' }}>{JSON.stringify(hasExtendedLicense)}</span>
			</p>
			<div>
				{state.payWhatYouWant && <div>
          <span>PWYW enabled</span>
          <NumberDialStyled
            label='Pay what you want'
            qty={state.price}
            disabled={state.inCart}
            inputOnChange={onPwywChange}/>
        </div>}
			</div>
			<hr/>
			<p>
				<LicenseSelect
					onChange={selectChange}
					selectedLicense={state.selectedLicense}
					showDropdown={hasExtendedLicense}
					inCart={state.inCart}
				/>
			</p>
			<hr/>
			<div style={{ paddingBottom: 20 }}>
				{state.bulkDiscount && <span>Bulk discount of {config.siteMetadata.pricing.discount} applied</span>}
				{!state.payWhatYouWant && <NumberDialStyled
          label='Select number of license'
          qty={state.numberOfLicenses}
          disabled={state.inCart}
          inputOnChange={onDialChange}/>}
			</div>
			<hr/>
			<div>
				{React.useMemo(() => (
					<AddToCartBtn
						handleAddToCartState={() => (setState({ inCart: true }))}
						isInCart={state.inCart}
						slug={product.slug}
						selectedProduct={state.selectedProduct}
						licenseQty={state.numberOfLicenses}
						price={state.price}
					/>), [state.inCart, state.price])}
			</div>
			<Link to='/page-2/'>Go to page 2</Link>
		</Layout>
	)

}
const NumberDialStyled = styled(NumberDial)`
	input{
		color: red;
		-moz-appearance: textfield;
		&::-webkit-outer-spin-button, 
		::-webkit-inner-spin-button{
			-webkit-appearance: none;
			margin: 0;
		};
	}
`
const mapStateToProps = (state: IState) => ({
	cart: state.cart,
	products: state.products
})
export default connect<IPropsPrivate, {}, IPropsPublic, IState>(mapStateToProps)(ProductLayout)