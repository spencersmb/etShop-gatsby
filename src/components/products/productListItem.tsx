import { IProduct, IProductFeaturedImage } from '@et/types/Products'
import { device } from '@styles/global/breakpoints'
import { colors } from '@styles/global/colors'
import { Sentinel } from '@styles/global/fonts'
import { shadowStyles } from '@styles/global/shadows'
import { svgs } from '@svg'
import { renderSvg } from '@utils/styleUtils'
import { Link } from 'gatsby'
import Img from 'gatsby-image'
import React from 'react'
import styled from 'styled-components'

interface IProps {
	id: string
	slug: string
	name: string
	sub_header: string
	price: string
	featuredImage: IProductFeaturedImage
}

function ProductListItem (props: IProps) {
	const { id, slug, name, sub_header, price, featuredImage } = props
	return (
		<ListItem key={id} data-testid={'listItem'}>
			<Link to={`/products/${slug}`}>
				<ListItemInner>
					<ListItemTop className='ListItemTop'>
						<Img
							alt={featuredImage.alt}
							fluid={featuredImage.localFile.childImageSharp.fluid}
						/>
						<ListItemTitle>
							<h2>
								{name}
							</h2>
							{sub_header && <p>{sub_header}</p>}
						</ListItemTitle>
					</ListItemTop>
					<ListItemFooter className='ListItemFooter'>
						<span/>
						<p>
							${price}
						</p>
						{renderSvg(svgs.CardSmall)}
					</ListItemFooter>
				</ListItemInner>
			</Link>
		</ListItem>
	)
}

const ListItem = styled.div`
	grid-column: 2 / 4;
	border-radius: 15px 15px 0 0;
	box-shadow: 0 3px 4px rgba(161, 161, 161, 0.0);
	transition: .3s;
	display: flex;
	flex-direction: column;
	margin-bottom: 45px;
	
	@media ${device.tablet}{
		grid-column: span 6;
		margin-bottom: 0;
	}
	@media ${device.laptop}{
		grid-column: span 4;
	}
	
	@media ${device.laptopL}{
		grid-column: span 3;
			&:hover{
				box-shadow: ${shadowStyles.shadow4};
				transform: translateY(-10px);
				svg{
					filter: drop-shadow(0px 0 0 rgba(0, 0, 0, 0));
				}
				.ListItemTop{
					box-shadow: 0 0 0 rgba(161, 161, 161, 0.0);
				}
				.ListItemFooter{
					background: #e2e2e2;
				}
			}
	}
	
	a{
		display: flex;
		flex: 1;
	}
`
const ListItemInner = styled.div`
	display: flex;
	flex-direction: column;
	width: 100%;
	max-width: 400px;
	margin: 0 auto;
`
const ListItemTitle = styled.div`
	background: white;
	text-align: center;
	padding: 20px 20px 10px;
	flex: 1;
	display: flex;
	flex-direction: column;
	justify-content: center;
	h2{
		${Sentinel.reg};
		font-size: 23px;
		font-weight: bold;
		line-height: 24px;
		color: ${colors.grey.i800}
	}
	p{
		${Sentinel.reg};
		font-size: 16px;
		font-weight: 500;
		font-style: italic;
		color: ${colors.secondary.text};
		margin: 0;
	}
`
const ListItemTop = styled.div`
	overflow: hidden;
	transition: .3s;
	box-shadow: ${shadowStyles.shadow1};
	border-radius: 15px 15px 0 0;
	display: flex;
	flex-direction: column;
	height: 100%;
	align-items: stretch;
`
const ListItemFooter = styled.div`
		background: transparent;
		display: flex;
		transition: .3s;
		position: relative;
		align-items: flex-end;
		p{
			position: absolute;
			z-index: 2;
			left: 50%;
			top: 50%;
			transform: translateY(-50%) translateX(-50%);
			font-size: 21px;
			font-family: Sentinel, serif;
			font-weight: bold;
			color: ${colors.grey.i800}
		}
		span{
			background: white;
			height: 5px;
			width: 100%;
			position: absolute;
			top: -5px;
			left: 0;
			z-index: 1;
		}
		svg{
			transition: .3s;
			width: 100%;
			filter: drop-shadow(${shadowStyles.shadow1});
		}
`
export default ProductListItem
