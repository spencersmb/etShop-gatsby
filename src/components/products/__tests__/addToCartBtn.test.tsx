import { AddToCartBtn } from '@components/products/addToCartBtn'
import { LicenseEnum } from '@et/types/Cart'
import { ProductKey, testCartEmpty, testProducts } from '@redux/reduxTestUtils'
import { calcTotalQtyPrice } from '@utils/priceUtils'
import React from 'react'
import renderer from 'react-test-renderer'
import {
	render,
	cleanup
} from 'react-testing-library'
import 'jest-dom/extend-expect'

afterEach(cleanup)

const propsDefault = {
	handleAddToCartState: jest.fn(),
	isInCart: false,
	slug: ProductKey.WatercolorStd,
	selectedProduct: testProducts[ProductKey.WatercolorStd],
	licenseQty: 0,
	price: testProducts[ProductKey.WatercolorStd].price,
	total: calcTotalQtyPrice(testProducts[ProductKey.WatercolorStd].price, 0),
	bulkDiscount: false,
	selectedLicense: LicenseEnum.standard,
	cart: testCartEmpty,
	addToCart: jest.fn(),
	cartToggle: jest.fn()
}

const propsCheckout = {
	handleAddToCartState: jest.fn(),
	isInCart: true,
	slug: ProductKey.WatercolorStd,
	selectedProduct: testProducts[ProductKey.WatercolorStd],
	licenseQty: 1,
	price: testProducts[ProductKey.WatercolorStd].price,
	bulkDiscount: false,
	selectedLicense: LicenseEnum.standard,
	cart: testCartEmpty,
	addToCart: jest.fn(),
	cartToggle: jest.fn(),
	total: calcTotalQtyPrice(testProducts[ProductKey.WatercolorStd].price, 1)

}

describe('AddToCart Button', () => {

	it('renders correctly', () => {
		const tree = renderer
			.create(
				<AddToCartBtn {...propsDefault}/>
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('Should render correct total', () => {
		const propsTotal = {
			...propsDefault,
			licenseQty: 1,
			total: calcTotalQtyPrice(testProducts[ProductKey.WatercolorStd].price, 1)
		}
		const modalRender = render(<AddToCartBtn {...propsTotal}/>)
		expect(modalRender.getByTestId('total').innerHTML).toEqual('$16')
	})

	it('Should render I want this button.', () => {
		const modalRender = render(<AddToCartBtn {...propsDefault}/>)
		expect(modalRender.getByTestId('addToCart').innerHTML).toEqual('I want this')
	})

	it('Should not be able call addToCart Action', () => {
		// we have licenses set to 0 by default
		const modalRender = render(<AddToCartBtn {...propsDefault}/>)
		const btn = modalRender.getByTestId('addToCart')
		btn.click()
		expect(propsDefault.addToCart).toHaveBeenCalledTimes(0)
	})

	it('Should call addToCart Action', () => {
		const enabled = {
			...propsDefault,
			licenseQty: 1
		}
		const modalRender = render(<AddToCartBtn {...enabled}/>)
		const btn = modalRender.getByTestId('addToCart')
		btn.click()
		expect(propsDefault.addToCart).toHaveBeenCalledTimes(1)
	})

	it('Should render checkout button', () => {
		const modalRender = render(<AddToCartBtn {...propsCheckout}/>)
		expect(modalRender.getByTestId('checkout').innerHTML).toEqual('Checkout')
	})

	it('Should call cartToggle Action', () => {
		const modalRender = render(<AddToCartBtn {...propsCheckout}/>)
		const btn = modalRender.getByTestId('checkout')
		btn.click()
		expect(propsCheckout.cartToggle).toHaveBeenCalledTimes(1)
	})

})
