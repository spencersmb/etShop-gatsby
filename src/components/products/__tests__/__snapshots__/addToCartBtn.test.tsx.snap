// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToCart Button renders correctly 1`] = `
<div
  className="sc-bxivhb bqNnwj checkoutWrapper"
>
  <div
    className="sc-ifAKCX etLTSi buttonWrapper"
  >
    <div
      className="sc-EHOje gyVRIK addToCart__total"
      data-testid="total"
    >
      0
    </div>
    <button
      className="sc-htpNat sc-bZQynM dgJtvg"
      color="#18C3C3"
      data-testid="addToCart"
      disabled={true}
      onClick={[Function]}
    >
      I want this
    </button>
  </div>
</div>
`;
